### ISSUE TITLE
Connecting to DocumentDB or Mongo in sandbox lambda fails, no timeout

### ISSUE THREAD

**Issue by mikemaccana** [arc-repos/architect#354]
**Describe the bug**
In architect `5.6.3`, connecting to DocumentDB / Mongo will cause an AWS Lambda to not respond (no timeout or errors are produced, just no response).

Since Amazon has announced DocumentDB (which uses Mongo DB clients) as their official JSON-type storage for AWS, I am attempting to use the MongoDB client to connect to a database:

**To Reproduce**
1. Install MongoDB locally
2. Confirm DB connection URL works
```javascript
const MongoClient = require('mongodb').MongoClient

;(async function(request, context) {
    await MongoClient.connect('mongodb://localhost')
    console.log(`I connected successfully!`)    
})();
```

3. Make the following minimal lambda:

```javascript
const MongoClient = require('mongodb').MongoClient

exports.handler = async function http(req) {
  console.log(`Got incoming request!`)

  // The following line, if uncommented, will stop even the console.log above from working
  // Instead the entire HTTP request will time out
  await MongoClient.connect('mongodb://localhost')

  return {
    headers: {'content-type': 'text/html; charset=utf8'},
    body: '<h1>Hello world!</h1>'
  }
}
```
**What happens**

Accessing the lambda that uses DocumentDB will cause Architect to not respond - **no timeout, no errors. The `console.log` is also not displayed.**

 - Note the URL works
 - Even if it didn't, there should be a timeout 
 - This works fine in a non-Architect lambda made on the Lambda console, Only Architect in sandbox is affected. I have not tried Architect in production AWS yet (need to get sandbox working first!)

**Expected behavior**
An HTTP response

**Desktop (please complete the following information):**
 - OS: Windows 10.0.18362.0

Though the above is reproducible on any system.

**Comment by ryanblock** [comment]
This issue needs a bit of clarification: is this related to `sandbox` / local env, or is this breaking in Lambdas + APIG that Arc provisions?

**Comment by mikemaccana** [comment]
Architect 5.6.3 sandbox is the only place it fails.

I've tried AWS proper (without Architect) and it's fine.  

 I hit it in the title.

**Comment by ryanblock** [comment]
So you did, apologies!

**Comment by mikemaccana** [comment]
Made a repo to reproduce: just install Mongo and `npx sandbox`: https://github.com/mikemaccana/documentdb-arc-test

**Comment by mikemaccana** [comment]
Also tested this in prod to see if it's just the sandbox. However when deploying with `npx deploy`:

```
npx deploy
        app ~ mongotest
     region ~ us-west-2
    profile ~ default
    version ~ 5.6.3

∙∙. Preparing 1 Function: Installing dependencies in 1 Functions
src\http\get-index ... mongotest-productiontestapp-get-index ... Creating
Found new resources to create! Creating them now...
https://vh1413crrh.execute-api.us-west-2.amazonaws.com/staging
https://xyb0s7pfxg.execute-api.us-west-2.amazonaws.com/production
```

The [production site](https://xyb0s7pfxg.execute-api.us-west-2.amazonaws.com/production) says:

> Cannot find module 'mongodb'

Despite [mongodb being in package.json](https://github.com/mikemaccana/documentdb-arc-test/blob/master/src/http/get-index/package.json) and the [deploy guide](https://arc.codes/reference/arc-deploy) saying it will intall everything in `package.json`

Is there something else I need to do to ensure modules get installed?

**Comment by ryanblock** [comment]
Nope, what if you load up that Lambda in the console and browse its files? Is it in there?

**Comment by mikemaccana** [comment]
Yep - the files are there.

```
ls .\src\http\get-index\node_modules\mongodb\


    Directory:
    C:\Users\mikem\OneDrive\Documents\appless\documentdb-arc-test\src\http\get-index\node_modules\mongodb

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----         11-Apr-19 10:57 PM                lib
-a----         26-Oct-85  9:15 AM         134197 HISTORY.md
-a----         26-Oct-85  9:15 AM           2277 index.js
-a----         26-Oct-85  9:15 AM          11323 LICENSE.md
-a----         26-Oct-85  9:15 AM           2021 package.json
-a----         26-Oct-85  9:15 AM          15318 README.md
```

**Comment by mikemaccana** [comment]
[Deleted previous response, just woke up here and misread consoel as in terminal not AWS console)

Yep the files are on the AWS Console:


![image](https://user-images.githubusercontent.com/172594/56027478-0a956900-5d0e-11e9-91a2-879934f3952b.png)

**Comment by mikemaccana** [comment]
OK back to the issue in the sandbox. The Architect sandbox also fails connect to a completely different DB.

Eg:

```javascript
const rethinkdb = require('rethinkdb')

exports.handler = async function http(req) {
  console.log(`Got incoming request!`)

  // The following line, if uncommented, will stop even the console.log above from working
  // Instead the entire HTTP request will time out
  await rethinkdb.connect({ host: 'localhost', port: 28015 })

  return {
    headers: {'content-type': 'text/html; charset=utf8'},
    body: '<h1>Hello world!</h1>'
  }
}

```

Will fail - no console.log - on Architect sandbox.

However the DB is running and that address works, as proven by:

```javascript
const rethinkdb = require('rethinkdb')

;(async function(request, context) {
    await rethinkdb.connect({ host: 'localhost', port: 28015 })
    console.log(`I connected successfully!`)    
})();
```

Prints '`I connected successfully`.

**Comment by mikemaccana** [comment]
OK getting closer to fixing this:
 
 - Closing the connection in the lambda makes it work
 - The process spawned by `spawn.js` will never exit if the connection remains open
 - In `spawn.js`, it's possible for a lambda to time out **and no exception will be sent to the user** as it looks like `child.kill()` does not trigger the `child.on('close'...` event

**Comment by mikemaccana** [comment]
Yep got it. There's an error in `spawn.js`. Ru nthe code below:

```javascript
let spawn = require('child_process').spawn
var verySlowNodeCode = "setTimeout(function(){console.log(`Finished`)},20000)"
var log = console.log.bind(console)

var child = spawn('node', ['-e', verySlowNodeCode], { 
    shell: true,
    cwd: 'C:\\Users\\mikem\\OneDrive\\Documents\\appless\\documentdb-arc-test\\src\\http\\get-index',
    env: {}
})


let stdout = ''
let stderr = ''
let timeout = 5000;
// bake a timeout
let to = setTimeout(function () {
    log(`Timed out after ${timeout} ms!`)
    timedout = true
    child.kill()
}, timeout)

child.stdout.on('data', data => {
    log(`Got output data ${data}`)
    // always capture data piped to stdout
    // python buffers so you might get everything despite our best efforts
    stdout += data
})

child.stderr.on('data', data => {
    log(`Got error data ${data}`)
    stderr += data
})

child.on('close', function done (code) {
    log(`Closed!`)
    // Output any console logging from the child process
    
})

child.on('exit', function done (code) {
    log(`Exited!`)
    // Output any console logging from the child process
})
```

**Comment by ryanblock** [comment]
This is a great find!

Sounds like it's keeping the connection open, but not actually doing anything, which is in turn not streaming events back, which is in turn not providing an opportunity to determine whether it should consider things timed out, yes? If so, we might want to cover our asses by adding some kind of heartbeat / timeout check.

As a side note that absolutely doesn't resolve this bug, but that may make this a little easier: you can set sandbox spawn process timeout with `SANDBOX_TIMEOUT` env var.

**Comment by mikemaccana** [comment]
Thanks! 

I've sent a PR attached to the issue above which fixes it. https://github.com/arc-repos/architect/issues/355

Meanwhile, there's a second issue: working out a way for spawn to exit when finishing processing even if there's a connection open (and this the event loop isn't open). I've asked on https://stackoverflow.com/questions/55656272/make-node-finish-when-done-processing-even-if-the-event-loop-isnt-empty about this.

**Comment by mikemaccana** [comment]
Since I've found both issues #355 (fixed) and #358 (yet to be fixed) I'm closing this issue down.
